diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3f6f6192..b4bd01c0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -105,6 +105,10 @@ target_link_libraries(${library_name}
   nav2_route::route_server_core
 )
 
+target_link_options(${library_name} PRIVATE
+  "LINKER:--as-needed"
+)
+
 # Causes the visibility macros to use dllexport rather than dllimport,
 # which is appropriate when building the dll but not consuming it.
 # TODO: Make this specific to this project (not rviz default plugins)
