diff --git a/launch/ros_topic_streamer.launch.py b/launch/ros_topic_streamer.launch.py
index a891b64..19df154 100644
--- a/launch/ros_topic_streamer.launch.py
+++ b/launch/ros_topic_streamer.launch.py
@@ -27,7 +27,7 @@
             ),
             ExecuteProcess(
                 condition=IfCondition(signalling),
-                cmd=["gst-webrtc-signalling-server"],
+                cmd=["@gst_plugins_rs@/bin/gst-webrtc-signalling-server"],
                 additional_env={
                     "WEBRTCSINK_SIGNALLING_SERVER_LOG": "warn",
                 },